refactor: un-exposing PXE from e2e tests

benesjan opened the pull request:
e2e tests should interact only with a wallet (and maybe a node) because PXE is becoming only a lib used by a wallet. This PR is a step in that direction.

# Followup work
1. Drop PXE JSON RPC server,
2. have Wallet instantiate PXE instead of having it passed in as an arg to a constructor,
3. revisit the PXE methods exposed on the TestWallet.

benesjan commented:
* **#17163** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17163?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/17163?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/aztec/src/cli/cmds/start_pxe.ts:
```diff
@@ -1,36 +1,21 @@
+export type { CliPXEOptions, PXEServiceConfig };
 
-export async function startPXE(
+export async function startPXEServicesGetWallet(
```
We now always just used the pxe to instantiate a test wallet so I just return a wallet from this function. This will get nuked in a followup PR when PXE will stop being a service.

benesjan commented on yarn-project/bot/src/factory.ts:
```diff
@@ -23,58 +20,23 @@ import { PrivateTokenContract } from '@aztec/noir-contracts.js/PrivateToken';
 
   constructor(
     private readonly config: BotConfig,
-    dependencies: { pxe?: PXE; nodeAdmin?: AztecNodeAdmin; node?: AztecNode },
-  ) {
-    if (config.flushSetupTransactions && !dependencies.nodeAdmin) {
-      throw new Error(
-        `Either a node admin client or node admin url must be provided if transaction flushing is requested`,
-      );
-    }
-    if (config.senderPrivateKey && config.senderPrivateKey.getValue() && !dependencies.node) {
-      throw new Error(
-        `Either a node client or node url must be provided for bridging L1 fee juice to deploy an account with private key`,
-      );
-    }
-    if (!dependencies.pxe && !config.pxeUrl) {
-      throw new Error(`Either a PXE client or a PXE URL must be provided`);
-    }
-
-    this.nodeAdmin = dependencies.nodeAdmin;
-
-    if (dependencies.pxe) {
-      this.log.info(`Using local PXE`);
-      this.pxe = dependencies.pxe;
-    } else {
-      this.log.info(`Using remote PXE at ${config.pxeUrl!}`);
-      this.pxe = createPXEClient(config.pxeUrl!, getVersions(), makeTracedFetch([1, 2, 3], false));
-    }
-
-    if (dependencies.node) {
-      this.node = dependencies.node;
-    } else {
-      this.node = createAztecNodeClient(config.nodeUrl!, getVersions(), makeTracedFetch([1, 2, 3], false));
-    }
-
-    this.wallet = new TestWallet(this.pxe, this.node);
-  }
```
This was all very messy and the functionality seems to have been completely unused as we always fed in the real instances.

benesjan commented on yarn-project/bot/src/runner.ts:
```diff
@@ -24,18 +22,13 @@ export class BotRunner implements BotRunnerApi, Traceable {
+    private readonly telemetry: TelemetryClient,
+    private readonly aztecNodeAdmin?: AztecNodeAdmin,
   ) {
-    this.tracer = dependencies.telemetry.getTracer('Bot');
-    if (!dependencies.node && !config.nodeUrl) {
-      throw new Error(`Missing node URL in config or dependencies`);
-    }
-    const versions = getVersions();
-    const fetch = makeTracedFetch([1, 2, 3], true);
-    this.node = dependencies.node ?? createAztecNodeClient(config.nodeUrl!, versions, fetch);
-    this.nodeAdmin =
-      dependencies.nodeAdmin ??
-      (config.nodeAdminUrl ? createAztecNodeAdminClient(config.nodeAdminUrl, versions, fetch) : undefined);
```
Same here. This seems to have been unused.

Thunkar replied:
I would ask @alexghr to make sure this is not used with certain configs in production

benesjan replied:
By dropping the ugly unreadable `dependencies` arg and having real types there, the type checker figured out that nowhere in the code we pass in those values as `undefined`.

So this shouldn't be a problem.

But agree that would be good to get Alex's feedback as it's strange that it was unused.

alexghr replied:
This code path used to run before v2 but in this #16962 I changed it to create a node client as part of `start_node` so I guess it's no longer used. This is ok to remove!

benesjan commented on yarn-project/end-to-end/src/composed/e2e_pxe.test.ts:
```diff
@@ -1,12 +0,0 @@
-  return pxe;
-};
-
-pxeTestSuite('e2e_pxe', setupEnv);
```
Since PXE is becoming library, this became pointless.

benesjan commented on yarn-project/end-to-end/src/e2e_account_contracts.test.ts:
```diff
@@ -56,10 +53,10 @@ const itShouldBehaveLikeAnAccountContract = (
+      wallet = new TestWalletInternals(wallet.getPxe(), aztecNode);
 
-      const accountManager = await AccountManager.create(wallet, pxe, secret, accountContract, salt);
+      const accountManager = await AccountManager.create(wallet, wallet.getPxe(), secret, accountContract, salt);
```
The `.getPxe()` getter is a temporary workaround that will get resolved once Wallet instantiates the PXE.

benesjan commented on yarn-project/end-to-end/src/bench/utils.ts:
```diff
@@ -150,32 +149,6 @@ export async function waitTxs(txs: SentTx[], context: EndToEndContext, txWaitOpt
- * @param startingBlock - First l2 block to process.
- * @returns The new PXE.
- */
-export async function createNewPXE(node: AztecNode, contract: BenchmarkingContract): Promise<PXEService> {
```
This was unused

benesjan commented on yarn-project/aztec/src/testing/aztec_cheat_codes.ts:
```diff
@@ -2,18 +2,18 @@ import { Fr } from '@aztec/foundation/fields';
+     * The test wallet or pxe to use for getting notes
      */
-    public pxe: PXE,
+    public testWalletOrPxe: { getNotes(filter: NotesFilter): Promise<UniqueNote[]> },
```
Using TestWallet here would make this unusable for externals as it's likely their wallet will not expose the getNotes method. Just having the method defined in the type makes this PR a non-breaking change. Everything else should not be facing externals.

Thunkar replied:
If this is unused, I would drop the `loadPrivate` method. That is a (testing) wallet utility

benesjan replied:
I am not sure if this is unused as this can be used by externals. Or do you expect externals to use our testing wallet?

Thunkar replied:
Or their own. But this is false advertisement, I would deprecate it even if they're using it (because it uses the wrong mental model)

benesjan replied:
I would say this being on a cheatcode class makes it ok to have this method. It might also be useful when debugging typescript tests. We should just instruct devs to not use it in production (implement a utility function for that if needed). WDYT?

benesjan commented on yarn-project/accounts/src/testing/index.ts:
```diff
@@ -45,10 +45,11 @@ export function getInitialTestAccountsData(): Promise<InitialAccountData[]> {
  */
-export async function getDeployedTestAccounts(pxe: PXE): Promise<InitialAccountData[]> {
+export async function getDeployedTestAccounts(testWalletOrPxe: { getPxe(): PXE } | PXE): Promise<InitialAccountData[]> {
+  const pxe = 'getPxe' in testWalletOrPxe ? testWalletOrPxe.getPxe() : testWalletOrPxe;
```
Needed to do this ugly hack because currently TestWallet doesn't correctly load the accounts that were registered in PXE before the Wallet was instantiated. This will be easy to tackle once Wallets instantiate PXE instead of having it passed in as an arg.

Thunkar replied:
This utility should just die. `InitialAccountData` contains everything the`TestWallt` needs to create the accounts. If they're not deployed, well, the consumer should do it.

benesjan replied:
I think it makes sense to address this in step 2 of followup work as then only wallet will have access to pxe making it a necessity that this is dropped.

Thunkar commented on yarn-project/aztec/src/cli/cmds/start_pxe.ts:
```diff
@@ -1,36 +1,21 @@
-export type { PXEServiceConfig, CliPXEOptions };
+export type { CliPXEOptions, PXEServiceConfig };
 
-export async function startPXE(
```
I would drop "pxe" from the name here and just go for `createWallet`

benesjan replied:
I plan on just nuking this utility in step 1 or 2 of followup work. All these utils will change once PXE is not a standalone service. Now it's actually starting a PXE service and createWallet would be quite misleading name for this reason:
<img width="645" height="154" alt="image" src="https://github.com/user-attachments/assets/9844e354-a8e1-4a41-a1f9-bdc09732f75c" />

Thunkar commented on yarn-project/end-to-end/src/e2e_crowdfunding_and_claim.test.ts:
```diff
@@ -97,7 +95,7 @@ describe('e2e_crowdfunding_and_claim', () => {
     );
     const crowdfundingInstance = await crowdfundingDeployment.getInstance();
-    await pxe.registerAccount(crowdfundingSecretKey, await computePartialAddress(crowdfundingInstance));
+    await wallet.registerAccount(crowdfundingSecretKey, await computePartialAddress(crowdfundingInstance));
```
I think it would be better to extend `registerContract` so it accepts keys

benesjan replied:
That's beyond the scope of this PR. The point of this PR is to just un-expose PXE from tests. The test wallet functions will be revisited in a followup work:

<img width="645" height="154" alt="image" src="https://github.com/user-attachments/assets/0d8baa1f-08e6-4f5f-b17a-a1469bc873dd" />

Thunkar commented on yarn-project/end-to-end/src/e2e_deploy_contract/deploy_test.ts:
```diff
@@ -75,7 +74,7 @@ export class DeployTest {
 
   async registerRandomAccount(): Promise<AztecAddress> {
-    const completeAddress = await this.pxe.registerAccount(Fr.random(), Fr.random());
+    const completeAddress = await this.wallet.registerAccount(Fr.random(), Fr.random());
```
I dislike that the testwallet is getting so many methods. I'd try to refactor this into wallet-only tests

benesjan replied:
Beyond the scope of this PR. You agreed last week on the process of:
1. Moving the functions onto TestWallet,
2. unexposing pxe from tests,
3. then revisiting the methods.

I think this is the only reasonable approach how to make this big pxe refactor reviewable.

Thunkar commented on yarn-project/end-to-end/src/fixtures/utils.ts:
```diff
@@ -148,23 +142,23 @@ export const setupL1Contracts = async (
- * @returns Private eXecution Environment (PXE), logger and teardown function.
+ * @returns A test wallet, logger and teardown function.
  */
-export async function setupPXEService(
```
I'd go with `createWallet`, the fact that it starts a PXE is internal

benesjan replied:
It's starting a pxe service which is an exposed service so would say that name would be misleading. Anyway makes sense to address in step 1 of followup work section.

Thunkar commented on yarn-project/test-wallet/src/wallet/test_wallet.ts:
```diff
@@ -176,4 +179,30 @@ export abstract class BaseTestWallet extends BaseWallet {
       return this.pxe.simulateTx(txRequest, true /* simulatePublic */, true, true, { contracts: contractOverrides });
     }
   }
+
```
These are way too much. Idk if now, but they need to go

benesjan replied:
See step 3:

<img width="645" height="154" alt="image" src="https://github.com/user-attachments/assets/0aa15c41-31a9-47d1-b738-e27a89631e69" />

benesjan committed:
refactor: un-exposing PXE from e2e tests

e2e tests should interact only with a wallet (and maybe a node) because PXE is becoming only a lib used by a wallet. This PR is a step in that direction.

# Followup work
1. Drop PXE JSON RPC server,
2. have Wallet instantiate PXE instead of having it passed in as an arg to a constructor,
3. revisit the PXE methods exposed on the TestWallet.
